### PR TITLE
add dependency table

### LINKED ISSUE #189: add table for cargo dependencies
Take [this cargo.toml](https://github.com/dyc3/opentogethertube/blob/master/Cargo.toml) at the root of the OTT repo, copy it over to this repo, and [use typst's toml function](https://typst.app/docs/reference/data-loading/toml/) to load it and render a table in `balancer-design.typ`

### PR DESCRIPTION
**Pull request by cjrkoa**
Fixes #189

### CODE REVIEW

**Comment by dyc3** [comment]
Is this done? Did you mean to make it a draft?

File: balancer-design.typ [hunk 1]
```diff
@@ -17,3 +17,46 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+
+#let dependencies = toml("Cargo.toml")
+
+#figure(
+	table(
+		[*Dependencies*],
+		[#dependencies.workspace]
+	),
+	caption: "External dependencies required by the balancer"
+)
```
Comment (dyc3) [inline balancer-design.typ:31]: This table is formatted really poorly, and Darian is gonna complain about it. It also includes other metadata from the top level of the cargo.toml, which is irrelevant. 

What I'm looking for is a table that looks something like this:

| Crate   | Version | Features  |
|---------|---------|-----------|
| anyhow  | 1       |           |
| figment | 0.10.10 | toml, env |
| ...     |         |           |

File: balancer-design.typ [hunk 2]
```diff
@@ -17,3 +17,46 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+
+#table(
+	[*External Dependencies*],
+	[*Package Name/Description*],
+	[anyhow: Error handling],
+	[async-trait: Support for asyncronous functions in traits],
```
Comment (dyc3) [inline balancer-design.typ:37]: You can work these descriptions into the other table if you really want to, but it's not necessary.

File: balancer-design.typ [hunk 3]
```diff
@@ -17,3 +17,46 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+== Dependencies
+
+#let dependencies = toml("Cargo.toml")
+
+#figure(
+	table(
+		[*Dependencies*],
+		[#dependencies.workspace]
+	),
+	caption: "External dependencies required by the balancer"
```
Comment (dyc3) [inline balancer-design.typ:30]: nit:
```suggestion
	caption: "External Rust dependencies for all workspace crates"
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: balancer-design.typ (hunk 1 above)
Comment (cjrkoa) [reply balancer-design.typ:31]: I'll figure out formatting after break

File: balancer-design.typ [hunk 4]
```diff
@@ -17,3 +17,44 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+			
+			for entry in value.features {
+				features += entry + ", "
+			}
+
+			row.push(features)
+			rows.push(row)
+		}
+		else {
+			rows.push((key, "", ""))
```
Comment (dyc3) [inline balancer-design.typ:46]: This doesn't display the git repo/commit in the version column, nor any features that are enabled for those deps.

File: balancer-design.typ [hunk 5]
```diff
@@ -17,3 +17,44 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+		if(type(value) == str) {
+			rows.push((key, value, ""))
+		} 
+		else if(type(value) == dictionary and value.keys().contains(version)) {
+			let row = (key, value.version,)
+			let features = ""
+			
+			for entry in value.features {
+				features += entry + ", "
+			}
```
Comment (dyc3) [inline balancer-design.typ:40]: nit: this should be equivalent
```typst
let features = value.features.join(", ")
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: balancer-design.typ [hunk 6]
```diff
@@ -17,3 +17,43 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+		} 
+		else if(type(value) == dictionary and value.keys().contains(version)) {
+			let row = (key, value.version,)
+			let features = value.features.join(", ")
+
+			row.push(features)
+			rows.push(row)
+		}
+		else if(type(value) == dictionary and value.keys().contains(path)){
+			rows.push((key, value.path, ""))
```
Comment (dyc3) [inline balancer-design.typ:42]: This still doesn't show features.

**Review by dyc3 — CHANGES_REQUESTED** [review]
Please be a little more diligent.

File: Cargo.toml
Comment (dyc3) [inline Cargo.toml:1]: You can't modify the Cargo.toml file. The whole point is for this to be automatically generated from what is actually in the OTT repo. If we need to update this in the future, it needs to be as simple as literally copying the file from the OTT repo with no modifications.

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: balancer-design.typ [hunk 7]
```diff
@@ -17,3 +17,50 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+			if(value.keys().contains("version")) { row.push(value.version) }
+			else if(value.keys().contains("path")) { row.push(value.path) }
+			else if(value.keys().contains("git")) { row.push(value.git) }
+
+			if(value.keys().contains("features")) {
+				let features = value.features.join(", ")
+
+				row.push(features)
+			}
+			else if(value.keys().contains("rev")) { row.push(value.rev) }
```
Comment (dyc3) [inline balancer-design.typ:46]: `rev` is part of the version, its not a crate feature. A crate that uses `git` and `rev` to specify the version can also have `features`

File: balancer-design.typ [hunk 8]
```diff
@@ -17,3 +17,50 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+
+	for (key, value) in data {
+		if(type(value) == str) {
+			rows.push((key, value, ""))
+		} 
+		else if(type(value) == dictionary) {
+			let row = (key,)
+			
+			if(value.keys().contains("version")) { row.push(value.version) }
+			else if(value.keys().contains("path")) { row.push(value.path) }
```
Comment (dyc3) [inline balancer-design.typ:38]: The caption of the table states "External dependencies". Dependencies with `path` are internal dependencies. They should be excluded.

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: balancer-design.typ [hunk 9]
```diff
@@ -36,14 +36,20 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
 			
 			if(value.keys().contains("version")) { row.push(value.version) }
 			else if(value.keys().contains("path")) { row.push(value.path) }
-			else if(value.keys().contains("git")) { row.push(value.git) }
+			else if(value.keys().contains("git") and value.keys().contains("rev")) { 
+				row.push(
+					"git: " + value.git + "\n" +
+					"rev: " + value.rev
```
Comment (dyc3) [inline balancer-design.typ:42]: nit: just render this as content, no need for it to be a string

File: balancer-design.typ [hunk 10]
```diff
@@ -36,14 +36,20 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
...
+				) 
+			}
+			else if(value.keys().contains("git")) { row.push("git: " + value.git) }
+			else if(value.keys().contains("rev")) { row.push("rev: " + value.rev) }
```
Comment (dyc3) [inline balancer-design.typ:46]: nit: technically, `rev` is only valid if `git` is present, so this extra branch is unnecessary

File: balancer-design.typ [hunk 11]
```diff
@@ -62,5 +68,5 @@ In Rust, packages are called "crates". The Balancer and Harness is split into mu
 
 #figure(
 	build_table(crates),
-	caption: "External Rust dependencies required for all workspace crates"
+	caption: "Internal and External Rust dependencies required for all workspace crates"
```
Comment (dyc3) [inline balancer-design.typ:71]: nit: Literally not what I asked you to do, but whatever.

**Review by dyc3 — APPROVED** [review]

**Pull request merged**